libcodechecker/report.py

```diff
@@ -0,0 +1,444 @@
+
+    There are multiple ids to identify a report.
+
+    report_id
+      * Should be unique it is a combination of the bug hash and the path hash.
+
+    report_hash
+      * A unique identifier should be generate by the analyzer if not this
+        module generates an identifier based on the bug path if available.
+
+    obsolate_report_hash
+      * There is already an algorithm to generate a unique identifier which
+        should be changed and not used after v6.0.
+
+    path_hash
+      * Used to identify a bug path if the report_hash is not unique enough.
+      * It is possible that the same hash is generated for different reports
+        where the bug paths are different to identify these cases a
+        separate path hash is generated (starts in separate cpp files but
+        ends in the same header).
+    """
+
+    def __init__(self, checker_name, report_hash=None,
+                 category=None, type=None):
+
+        self.checker_name = checker_name
+        self.category = category
+        self.type = type
+        self.__hash_value = report_hash
+
+        # The main diag section for the report.
+        self.__main_section = None
+        # Diag sections without a path.
+        self.__diag_sections = []
+        # An optional bug path.
+        self.__bug_path = None
+
+    def diag_sections(self):
+        return self.__diag_sections
+
+    @property
+    def file_path(self):
+        self.__main_section.start_range.begin.file_path
+
+    @property
+    def main_section(self):
+        """
+        Main section after v6.0.
+        """
+        return self.__main_section
+
+    @main_section.setter
+    def main_section(self, main_section):
+        self.__main_section = main_section
+
+    @property
+    def obsolate_main_section(self):
+        """
+        Main section before v6.0.
+        """
+        return self.__bug_path.get_diag_sections()[-1]
+
+    def add_diag_section(self, section):
+        self.__diag_sections.append(section)
+
+    def set_path(self, bug_path):
+        self.__bug_path = bug_path
+
+    def get_path(self):
+        return self.__bug_path
+
+    @property
+    def path_hash(self):
+        if not self.__path_hash:
+            self.__path_hash = self.__bug_path.path_hash
+        return self.__path_hash
+
+    def report_id(self):
+        return self.hash_value + "||" + self.__bug_path.path_hash
+
+    @property
+    def report_hash(self):
+        """
+        If hash value is not set in the analyzers output generate a new.
+        """
+        if not self.__hash_value:
+            # NOTE  uses a different algorithm compared to obsolate hash.
+            self.__hash_value = self.__gen_report_hash()
+
+        return self.__hash_value
+
+    def events(self):
+        # TODO: FOR STORAGE BACKWARD COMPATIBILITY REMOVE AFTER v6.0
+        return [x for x in self.__bug_path.get_diag_sections() if
+                x.kind == "event"]
+
+    def paths(self):
```

Comment:
This will be removed in the upcoming refactoring, kept here for backward compatibility with the current way of storing the path elements.